rfcs/0008-microtvm-project-api.md

```diff
@@ -0,0 +1,509 @@
+   before all of `data` was sent. No timeout if `timeout_sec` is None.
+
+This is a departure from the previous interface, which allowed the implementation to read or write
+less data than was needed, returning as soon as possible. This was done initially to match the
+typical UNIX `read` and `write` semantics, but it turns out this was tricky to implement with a
+timeout. The reason for this was as follows:
+
+1. Because these semantics were expected from the interface, TVM always commanded it to read `128`
+   bytes, even if less were needed.
+2. However, not all libraries obeyed these semantics. In particular, PySerial reads data until the
+   timeout occurs, possibly returning less than `n` bytes. The implementation tried to read 1 byte
+   until `timeout_sec`, then set `timeout_sec` to the expected time taken to transmit `n - 1` bytes,
+   and returned whatever it could within that window.
+3. In the case where `timeout_sec` was `None` (e.g. when debugging something else), implementations
+   _should_ be quite simple. However, this wasn't the case, because TVM mostly requested more data
+   than it actually needed. In this case, implementations using PySerial were forced to return 1
+   byte, causing a lot of log spam given the number of round-trips. Also, implementations using
+   file descriptors were needlessly complicated, since `select` plus a non-blocking read was needed.
+
+In the new interface, implementers know exactly how much data to read and write, and the deadline
+for such operations. This interface is overall easier to implement and aligns it better with
+PySerial. Implementations can choose the simplest approach, which is particularly beneficial given
+that it will be more brittle to depend on shared Python modules from API Server implementations.
+
+# Prior art
+[prior-art]: #prior-art
+
+The current way of integrating with third-party platforms is via the abstractions in the `tvm.micro`
+namespace:
+
+1. `tvm.micro.Compiler`: used to produce binary and library artifacts.
+2. `tvm.micro.Flasher`: used to program attached hardware
+3. `tvm.micro.Transport`: used to communicate with on-device microTVM RPC server
+
+There are several drawbacks to the present scenario:
+
+1. Generally speaking, this interface encourages code from any platform used by TVM to live in the
+   TVM tree (the main barrier is the CI and reviews).
+2. The `Compiler` interface is not the correct abstraction of a platform's build system.
+3. The interface is large and spread across multiple classes, making it difficult to assemble a
+   list of tasks needed to support new platforms.
+4. The implementation is overly complex, making it diffcult to actually support such platforms.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+1. Is anyone particularly opposed the RPC mechanism used here?
+2. Does this seem simple for downstream platforms to implement?
```

Comment:
Do you have an example implementation?

Reply:
a simple example is the subprocess-based microtvm test: https://github.com/areusch/incubator-tvm/tree/project-generator/src/runtime/crt/host

more complex is the zephyr implementation: https://github.com/areusch/incubator-tvm/tree/project-generator/apps/microtvm/zephyr/template_project

i'd recommend reading the first one if you're not familiar with zephyr; the second one is not a minimal example and includes code to communicate with qemu and physical boards